lib/call-cabal-project-to-nix.nix

```diff
@@ -94,21 +101,30 @@ let
           (builtins.head pair)
           (builtins.elemAt pair 1))) blockLines);
 
-  fetchRepo = repo: (pkgs.fetchgit {
-    url = repo.location;
-    rev = repo.tag;
-    sha256 = repo."--sha256";
-  }) + (if repo.subdir or "" == "" then "" else "/" + repo.subdir);
+  # Use pkgs.fetchgit if we have a sha256. Add comment like this
+  #   --shar256: 003lm3pm0000hbfmii7xcdd9v20000flxf7gdl2pyxia7p014i8z
+  # otherwise use __fetchGit.
+  fetchRepo = repo:
+    let sha256 = repo."--sha256" or (lookupSha256 repo);
+    in (if sha256 != null
+      then pkgs.fetchgit {
+          url = repo.location;
+          rev = repo.tag;
+          inherit sha256;
+        }
+      else builtins.fetchGit {
```

Comment:
I think this case is worthy of a `trace` warning since it's an easy mistake to make and will now be much more likely to "work" and then break when they do it on Hydra.

Reply:
Good call!  We can even suggest the hash value.  Output looks like this when the DL works:

```
$ nix-build -A cabal-source-repo
trace: WARNING: No sha256 found for source-repository-package https://github.com/input-output-hk/haskell.nix.git bc01ebc05a8105035c9449943046b46c8364b932 download may fail in restricted mode (hydra)
trace: Consider adding `--sha256: 003lm3pm024vhbfmii7xcdd9v2rczpflxf7gdl2pyxia7p014i8z` to the cabal.project file or passing in a lookupSha256 argument
```

And like this when it fails:

```
$ nix-build --option restrict-eval true -I .. --option allowed-uris '[ https://github.com/NixOS/nixpkgs/archive https://github.com/input-output-hk/hackage.nix ]' -A cabal-source-repo
trace: WARNING: No sha256 found for source-repository-package https://github.com/input-output-hk/haskell.nix.git bc01ebc05a8105035c9449943046b46c8364b932 download may fail in restricted mode (hydra)
error: access to URI 'https://github.com/input-output-hk/haskell.nix.git' is forbidden in restricted mode
(use '--show-trace' to show detailed location information)
```